File: app/src/main/java/com/anokmik/tripassistant/trip/details/TripDetailsPresenter.java
package com.anokmik.tripassistant.trip.details;

import android.databinding.ObservableBoolean;

import com.anokmik.persistence.model.Trip;
import com.anokmik.persistence.model.TripEvent;
import com.anokmik.persistence.model.TripEvent_Table;
import com.anokmik.persistence.model.Trip_Table;
import com.anokmik.persistence.repository.TripEventRepository;
import com.anokmik.persistence.repository.TripRepository;
import com.anokmik.tripassistant.databinding.ObservableCompositeList;
import com.anokmik.tripassistant.databinding.adapter.OnItemClickListener;
import com.anokmik.tripassistant.databinding.adapter.ViewHolderPresenter;
import com.anokmik.tripassistant.model.ObservableTrip;
import com.anokmik.tripassistant.trip.Mode;
import com.anokmik.tripassistant.util.DateUtils;
import com.anokmik.tripassistant.validator.TextLengthValidator;

import java.util.List;

public final class TripDetailsPresenter implements TripDetailsContract.Presenter, TripDetailsContract.TripEventListener, OnItemClickListener<TripEvent> {

    public final ObservableTrip observableTrip;
    public final ObservableBoolean isEditing;
    public final ObservableBoolean titleValid;

    private final TripRepository tripRepository;
    private final TextLengthValidator validator;
    private final TripDetailsContract.View view;
    private final int mode;
    private final long tripId;
    private final ObservableCompositeList<TripEvent> tripEvents;

    public TripDetailsPresenter(TripDetailsContract.View view, @Mode int mode, long tripId) {
        this.tripRepository = new TripRepository();
        this.validator = new TextLengthValidator();
        this.view = view;
        this.mode = mode;
        this.tripId = tripId;
        this.observableTrip = new ObservableTrip(getTrip());
        this.tripEvents = new ObservableCompositeList<>(new TripEventRepository().getAsyncList(TripEvent_Table.trip.is(tripId)));

        this.isEditing = new ObservableBoolean(isEditingInitialValue());
        this.titleValid = new ObservableBoolean(titleValidInitialValue());
    }

    @Override
    public List<TripEvent> getTripEvents() {
        return tripEvents;
    }

    @Override
    public ViewHolderPresenter<TripEvent> getViewHolderPresenter() {
        return new ViewHolderPresenter.Builder<TripEvent>(view.getRowItemLayoutId(), view.getItemBindingId())
                .setAdapterPositionProviderBindingId(view.getAdapterPositionProviderBindingId())
                .setItemClickListener(this)
                .mapVariable(view.getItemListenerBindingId(), this)
                .mapVariable(view.getItemIsEditingBindingId(), isEditing)
                .build();
    }

    @Override
    public void showStartDatePicker() {
        view.showStartDatePickerDialog(observableTrip.getStartDate());
    }

    @Override
    public void showFinishDatePicker() {
        view.showFinishDatePickerDialog(observableTrip.getFinishDate());
    }

    @Override
    public void setStartDate(long startDate) {
        observableTrip.setStartDate(startDate);
    }

    @Override
    public void setFinishDate(long finishDate) {
        observableTrip.setFinishDate(finishDate);
    }

    @Override
    public boolean validFields() {
        titleValid.set(validator.notEmpty(observableTrip.getTitle()));
        boolean validDates = DateUtils.validDates(observableTrip.getStartDate(), observableTrip.getFinishDate());
        if (!validDates) {
            view.showDatesInvalidError();
        }
        return titleValid.get() && validDates;
    }

    @Override
    public void save() {
        if (validFields()) {
            isEditing.set(false);
            observableTrip.save();
            handleSave();
        }
    }

    @Override
    public void cancel() {
        handleCancel();
    }

    @Override
    public void edit() {
        isEditing.set(true);
        view.enableSaveControls();
    }

    @Override
    public void delete() {
        for (TripEvent tripEvent : getTripEvents()) {
            tripEvent.delete();
        }
        observableTrip.delete();
        view.back();
    }

    @Override
    public void addEvent() {
        view.addTripEvent();
    }

    @Override
    public void deleteEvent(int position) {
        tripEvents.remove(position);
    }

    @Override
    public void onItemClick(TripEvent item, int position) {
        view.viewTripEvent(item.id);
    }

    private boolean isEditingInitialValue() {
        switch (mode) {
            case Mode.ADD:
                return true;
            case Mode.VIEW:
                return false;
            default:
                throw new IllegalArgumentException();
        }
    }

    private boolean titleValidInitialValue() {
        switch (mode) {
            case Mode.ADD:
            case Mode.VIEW:
                return true;
            default:
                throw new IllegalArgumentException();
        }
    }

    private Trip getTrip() {
        switch (mode) {
            case Mode.ADD:
                Trip trip = new Trip();
                trip.startDate = System.currentTimeMillis();
                trip.finishDate = System.currentTimeMillis();
                return trip;
            case Mode.VIEW:
                return tripRepository.get(Trip_Table.id.is(tripId));
            default:
                throw new IllegalArgumentException();
        }
    }

    private void handleSave() {
        switch (mode) {
            case Mode.ADD:
                view.back();
                break;
            case Mode.VIEW:
                view.enableEditControls();
                break;
            default:
                throw new IllegalArgumentException();
        }
    }

    private void handleCancel() {
        switch (mode) {
            case Mode.ADD:
                view.back();
                break;
            case Mode.VIEW:
                isEditing.set(false);
                view.enableEditControls();
                observableTrip.set(getTrip());
                break;
            default:
                throw new IllegalArgumentException();
        }
    }

}

File: app/src/main/kotlin/com/anokmik/tripassistant/trip/details/TripDetailsPresenter.kt
package com.anokmik.tripassistant.trip.details

import android.databinding.ObservableBoolean
import com.anokmik.persistence.model.Trip
import com.anokmik.persistence.model.TripEvent
import com.anokmik.persistence.model.Trip_Table
import com.anokmik.persistence.model.TripEvent_Table
import com.anokmik.persistence.repository.TripEventRepository
import com.anokmik.persistence.repository.TripRepository
import com.anokmik.tripassistant.databinding.ObservableCompositeList
import com.anokmik.tripassistant.databinding.adapter.ViewHolderPresenter
import com.anokmik.tripassistant.model.ObservableTrip
import com.anokmik.tripassistant.trip.ADD
import com.anokmik.tripassistant.trip.Mode
import com.anokmik.tripassistant.trip.VIEW
import com.anokmik.tripassistant.util.DateUtils
import com.anokmik.tripassistant.validator.TextLengthValidator

class TripDetailsPresenter(private val view: TripDetailsContract.View, @Mode val mode: Long, private val tripId: Long) : TripDetailsContract.Presenter {

    val isEditing = ObservableBoolean(isEditingInitialValue())
    val titleValid = ObservableBoolean(titleValidInitialValue())

    private val tripRepository = TripRepository()
    private val tripEventRepository = TripEventRepository()
    private val validator = TextLengthValidator()

    private val trip: Trip?
        get() {
            when (mode) {
                ADD -> {
                    val trip = Trip()
                    trip.startDate = System.currentTimeMillis()
                    trip.finishDate = System.currentTimeMillis()
                    return trip
                }
                VIEW -> return tripRepository.get(Trip_Table.id.`is`(tripId))
                else -> throw IllegalArgumentException()
            }
        }

    val observableTrip = ObservableTrip(trip)

    override val tripEvents = ObservableCompositeList(tripEventRepository.getAsyncList(TripEvent_Table.trip.`is`(tripId)))

    override val viewHolderPresenter = ViewHolderPresenter.Builder<TripEvent>(view.rowItemLayoutId, view.itemBindingId)
            .setItemClickListener(this)
            .setAdapterPositionProviderBindingId(view.adapterPositionProviderBindingId)
            .mapVariable(view.itemListenerBindingId, this)
            .mapVariable(view.itemIsEditingBindingId, isEditing)
            .build()

    override fun showStartDatePicker() = view.showStartDatePickerDialog(observableTrip.startDate)

    override fun showFinishDatePicker() = view.showFinishDatePickerDialog(observableTrip.finishDate)

    override fun setStartDate(startDate: Long) {
        trip?.startDate = startDate
    }

    override fun setFinishDate(finishDate: Long) {
        trip?.finishDate = finishDate
    }

    override fun validFields(): Boolean {
        titleValid.set(validator.notEmpty(observableTrip.title))
        val validDates = DateUtils.validDates(observableTrip.startDate, observableTrip.finishDate)
        if (!validDates) {
            view.showDatesInvalidError()
        }
        return titleValid.get() && validDates
    }

    override fun save() {
        if (validFields()) {
            isEditing.set(false)
            observableTrip.save()
            handleSave()
        }
    }

    override fun cancel() = handleCancel()

    override fun edit() {
        isEditing.set(true)
        view.enableSaveControls()
    }

    override fun delete() {
        for (tripEvent in tripEvents) {
            tripEvent.delete()
        }
        observableTrip.delete()
        view.back()
    }

    override fun addEvent() = view.addTripEvent()

    override fun deleteEvent(position: Int) {
        tripEvents.remove(tripEvents[position])
    }

    override fun onItemClick(item: TripEvent?, position: Int) {
        item?.let {
            view.viewTripEvent(item.id)
        }
    }

    private fun isEditingInitialValue(): Boolean {
        when (mode) {
            ADD -> return true
            VIEW -> return false
            else -> throw IllegalArgumentException()
        }
    }

    private fun titleValidInitialValue(): Boolean {
        when (mode) {
            ADD, VIEW -> return true
            else -> throw IllegalArgumentException()
        }
    }

    private fun handleSave() {
        when (mode) {
            ADD -> view.back()
            VIEW -> view.enableEditControls()
            else -> throw IllegalArgumentException()
        }
    }

    private fun handleCancel() {
        when (mode) {
            ADD -> view.back()
            VIEW -> {
                isEditing.set(false);
                view.enableEditControls();
                observableTrip.set(trip);
            }
            else -> throw IllegalArgumentException()
        }
    }

}